### PR TITLE
[FLINK-13440] Add test that fails job when in-flight sync savepoint is discarded.

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

This PR adds a test for the changes introduced in https://github.com/apache/flink/pull/9131. This PR is a follow-up of https://github.com/apache/flink/pull/9131 and also contains those commits.

## Verifying this change

This change added tests and can be verified by running the `CheckpointCoordinatorTest.jobFailsIfInFlightSynchronousSavepointIsDiscarded()`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureReason.java [hunk 1]
```diff
@@ -23,60 +23,69 @@
...
 
 	public String message() {
 		return message;
 	}
+
+	/**
+	 * @return true if this value indicates a failure reason happening before a checkpoint is passed to a job's tasks.
+	 */
+	public boolean isPreFlight() {
+		return isPreFlight;
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureReason.java]: consider renaming the field to `preFlight`: https://stackoverflow.com/questions/5322648/for-a-boolean-field-what-is-the-naming-convention-for-its-getter-setter

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java [hunk 2]
```diff
@@ -3994,6 +4008,130 @@ public void testSharedStateRegistrationOnRestore() throws Exception {
...
+
+		final AtomicReference<Throwable> actualCause = new AtomicReference<>();
+		savepointFuture.handle((path, throwable) -> {
+			if (throwable != null) {
+				actualCause.set(throwable);
+			}
+			return path;
+		}).get();
+
+		assertThat(actualCause.get(), equals(expectedRootCause));
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java]: Consider `ExceptionUtils#stripCompletionException()` and `ExceptionUtils#findThrowableWithMessage()`

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java [hunk 3]
```diff
@@ -120,13 +122,21 @@
...
+	private Throwable failureCause;
+	private int failureCallbackInvocationCounter = 0;
 
 	@Rule
 	public TemporaryFolder tmpFolder = new TemporaryFolder();
 
 	@Before
 	public void setUp() throws Exception {
-		failureManager = new CheckpointFailureManager(0, () -> {});
+		failureCallbackInvocationCounter = 0;
```
Comment (aljoscha) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java]: This new code is specific to only one test, I would suggest to create new `CheckpointFailureManager` inside the test to isolate those changes in that test alone.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java [hunk 4]
```diff
@@ -1229,7 +1229,8 @@ private boolean processFail(Throwable t, boolean isCallback, Map<String, Accumul
 		}
 	}
 
-	boolean switchToRunning() {
+	@VisibleForTesting
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java]: This is not needed after the test refactoring, right?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 5]
```diff
@@ -581,10 +580,10 @@ public void enableCheckpointing(
 
 		checkpointStatsTracker = checkNotNull(statsTracker, "CheckpointStatsTracker");
 
-		CheckpointFailureManager failureManager = new CheckpointFailureManager(chkConfig.getTolerableCheckpointFailureNumber(), () ->
-			getJobMasterMainThreadExecutor().execute(() ->
-				failGlobal(new FlinkRuntimeException("Exceeded checkpoint tolerable failure threshold."))
-			));
+		CheckpointFailureManager failureManager = new CheckpointFailureManager(
+				chkConfig.getTolerableCheckpointFailureNumber(),
+				cause -> getJobMasterMainThreadExecutor().execute(() -> failGlobal(cause))
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: Do we now not have this error message anywhere anymore?

Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java]: The previous message is just moved "upwards" in the `CheckpointFailureManager.handleCheckpointException()`.

**Comment by kl0u** [comment]
I integrated your comments @aljoscha . Please have a look and let me know.

**Comment by kl0u** [comment]
OK I will merge.

**Comment by kl0u** [comment]
Merged

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureReason.java [hunk 6]
```diff
@@ -23,60 +23,69 @@
...
+	CHECKPOINT_DECLINED_ON_CANCELLATION_BARRIER(false, "Task received cancellation from one of its inputs"),
 
-	CHECKPOINT_DECLINED_ALIGNMENT_LIMIT_EXCEEDED("The checkpoint alignment phase needed to buffer more than the configured maximum bytes"),
+	CHECKPOINT_DECLINED_ALIGNMENT_LIMIT_EXCEEDED(false, "The checkpoint alignment phase needed to buffer more than the configured maximum bytes"),
 
-	CHECKPOINT_DECLINED_INPUT_END_OF_STREAM("Checkpoint was declined because one input stream is finished"),
+	CHECKPOINT_DECLINED_INPUT_END_OF_STREAM(false, "Checkpoint was declined because one input stream is finished"),
 
-	CHECKPOINT_COORDINATOR_SHUTDOWN("CheckpointCoordinator shutdown."),
+	CHECKPOINT_COORDINATOR_SHUTDOWN(false, "CheckpointCoordinator shutdown."),
```
Comment (klion26) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureReason.java]: https://github.com/apache/flink/blob/94fa4ceade57172362e2d35e5aac8383f8f40a40/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java#L472

Why the `preFlight` status of `CHECKPOINT_COORDINATOR_SHUTDOWN` is false? In `CheckpointCoordinator#triggerCheckpoint`  we'll throw `CHECKPOINT_COORDINATOR_SHUTDOWN` exception, seems the checkpoint has not passed into task already

**Pull request closed without merging**